### DISCUSSION TITLE
Lexicon v2 Progress Report: Beta Features, Branches, and Future Roadmap

### DISCUSSION THREAD

**Discussion by starjustice** [lexiconhq/lexicon#37] [General]
This post aims to provide an update on the current state of Lexicon. The Lexicon team is currently engaged in developing the beta version of Lexicon v2. In this upcoming release, the introduction of the Lexicon plugin will bring forth exciting new features including push notifications and email deep linking. These additions will seamlessly integrate with the Lexicon app, elevating its functionality. Additionally, the new version will include support for BBCode emojis, app linking from post or topic links, and address various bug fixes related to topic and post pagination.

In the Lexicon repository at [https://github.com/lexiconhq/lexicon](https://github.com/lexiconhq/lexicon), we currently have two branches: `master` and `alpha-v2`. The master branch holds our latest release for version 1 of Lexicon. However, please note that this version may not be entirely stable and may not work properly with the Expo Go app due to its utilization of an older Expo version. We would welcome PRs to make the master branch stable

On the other hand, the alpha-v2 branch is dedicated to the latest updates for Lexicon version 2. This branch includes a set of upcoming changes specifically designed for the beta v2 launch. In this update, we have upgraded Expo to a newer version and made necessary adjustments to other packages to ensure compatibility with Expo Go. Additionally, we have made significant performance improvements to the app, particularly when scrolling through topic and post data.

> **Please note that `master` and `alpha-v2` are intentionally incompatible.** Since `alpha-v2` represents a hard major version break from version 1 to version 2, it contains changes that would break on the original version 1 (`master`). Pull requests that attempt to merge all of `alpha-v2` into `master` cannot be accepted.
>
>
>
> However, it may be worth pulling some of the bugfixes and improvements from `alpha-v2` into `master` as issues are discovered.
>
>
>
> The important point to keep in mind is that the two branches diverge from one another and need to be maintained independently of one another.

In terms of our current approach and future plans, we are currently in the internal development phase for v2 of Lexicon. Once we have completed the beta launch and are ready to proceed, our intention is to transition to a fully open-source approach, operating the project solely on `lexiconhq/lexicon`. Until that point, we will periodically synchronize the changes made in our internal development repository with the public repository at `lexiconhq/lexicon` on the `alpha-v2` branch.

**Comment by SimonBiggs** [comment]
Hi @starjustice,

Thank you for this overview. It helps address my significant confusion :).

I'm keen to not reinvent the wheel, so essentially, until you make your internal branch open source I will opt to not help at all, so that I am not developing something that will, in the end, be difficult to reconcile with the pathway you are taking.

**Reply by Kinostrome** [reply]
Hey @SimonBiggs. Sure, that seems reasonable. 

However, we wouldn't want that message to discourage anyone from contributing to version 1 on the `master` branch to address any current issues that they might be encountering.

For example, we're preparing to merge in one of the open PRs related to upgrading Expo, which may address some of the problems that could be occurring on that branch.

Similar fixes or adjustments would also be welcome as PRs to be reviewed and merged. It'd be great to know what other issues people are encountering with version 1 at the moment.

On the other hand, large sets of changes or major refactors may not be as productive.

In short, there are still plenty of ways to contribute to version 1. And we'd encourage everyone to keep trying out the `alpha-v2` branch as well.

---

By the way, on this point...

> until you make your internal branch open source

Just to clarify, our internal progress is periodically synced to a public **branch** on this repository—`alpha-v2`. The aspect of this point which is internal is actually the _repository_ itself within our private organization. This is what will later be open-sourced (by simply moving all development to this repository at lexiconhq/lexicon rather than our internal one).

**Comment by gabriellaputri** [comment]
Friday, July 21, 2023

Synced changes to `alpha-v2` branch:
- Show a detailed error message in a popup when opening a private message
- Update to Expo 49
- Bump semver from 5.7.1 to 5.7.2 in /documentation
- Bump semver from 5.7.1 to 5.7.2 in /api
- Bump word-wrap from 1.2.3 to 1.2.4 in /api
- Bump tough-cookie from 4.1.2 to 4.1.3 in /api
- Bump word-wrap from 1.2.3 to 1.2.4 in /frontend
- Bump semver from 5.7.1 to 5.7.2
- Bump json5 from 1.0.1 to 1.0.2 in /frontend
- Bump semver from 5.7.1 to 5.7.2 in /frontend

**Comment by Bardreamaster** [comment]
I'm keep following the v2 progress. And the latest time I have tried v2 branch is on Jul 6. But I remember I run fronted and api without error report but on iphone expo report connection to server error. I just thought maybe something needs to be added. So I always check you commits and roadmap.

Back to important thing: Will there be an official release/public on GitHub Dissusion or Discourse Forum to show Lexicon-v2 is ready to deploy for users?   I'm really looking forward it.

And maybe there are something I (who is not good at web develop but other kind of software developer using Lexicon) can do in the future (maybe like translate Docs or what). HHhhhh.

**Comment by jovitazhang** [comment]
Hi @Bardreamaster, regarding your issues:
It is possible there is a bug. Could you file a detailed bug report with screenshots in it? Therefore, we could check it more further. In addition, you probably could setup the prose server to  see if you have missing setup steps there. 

Yes, we are currently working toward our upcoming version 2 beta release, and this announcement will be posted on our social media. You can  also check the update there.

On the other hand, we'd greatly appreciate translations of the documentation if you interested in doing that. it will really helpful.

Afterall, thanks for reaching this out to us and we will be happy to help!

**Reply by Bardreamaster** [reply]
@jovitazhang I'm trying to deploy the v2 recently using the latest codes. But I have few questions about deployment:

- I've noticed some documents have been updated, should I follow the v2 document locally to run v2 quick-start demo？ (If I still meet some error can not be handled by the latest, I'll ask on another new topic/issue)

And few about project itself:

- I may could translate the docs after v2-beta released. If any workflow is recommended, please let me know. (I've used docusaurus-v2)
- I don't know about your any social media. If possible please let me know exactly.

Thanks!

**Reply by jovitazhang** [reply]
@Bardreamaster It seems like a correct one. But if I may ask, could you please provide the documents link that you're referring to? so that we won't give an incorrect information. 

Regarding the workflow, we don't have any recommendation for now but we could discuss after you make a PR using the docusaurus -v2


Furthermore, for our social media :
- ProductHunt Page: https://kfox.io/lexph
- Marketing Site: https://lexicon.is/
- Instagram: https://www.instagram.com/getlexicon/
- Twitter: https://twitter.com/getlexicon

Thank you!

**Reply by Bardreamaster** [reply]
@jovitazhang  I'm referring to `http://localhost:3001/next/quick-start`, the next version docs.

Thanks for help.

**Reply by jovitazhang** [reply]
@Bardreamaster after we checked the link that you are referring to, It's a correct one. 
Please let us know if you have more question. Thank you

**Comment by gabriellaputri** [comment]
Thursday, July 27, 2023

Synced changes to `alpha-v2` branch:
- Add `yarn generate` command in dockerfile

**Comment by gabriellaputri** [comment]
Thursday, August 3, 2023

Synced changes to `alpha-v2` branch:
- Fix bug when redirecting deep link after login and implement new condition for deep link navigation
- Fix authentication bugs that occur after refactoring deep linking
- Fix navigation race condition between `instanceLoading` and `getStateFromPath`
- Fix warning package and adjust constant for api
- Setting up a logger for the Prose server using `winston`

**Comment by gabriellaputri** [comment]
Friday, August 11, 2023

Synced changes to `alpha-v2` branch:
- Add versioned docs and sidebar for v2.0.0-beta
- Improvement on the logging system:
   - Add info in the log about which Prose API triggers the error
   - Utilize GraphQL Yoga plugin to log errors from one place

**Comment by gabriellaputri** [comment]
Wednesday, August 16, 2023

Synced changes to `alpha-v2` branch:
- Updating package `expo-file-system` and `expo-update` version
- Undo certain word changes within the API's error messages
- Fix issue from https://github.com/lexiconhq/lexicon/discussions/43
   - Normalize `description` from `posters` across multiple languages by mapping each string in the locale of interest to a well-defined "symbol"

**Comment by Kinostrome** [comment]
Friday, August 18, 2023

Synced changes to alpha-v2 branch (performed by @starjustice):

- sync deleted files to `alpha-v2`
   - delete documentation about the lexicon plugin in doc version 1.0.0 
   - delete patch react-native 0.69.6
   - delete stack navigator `frontend/src/navigation/StackNavigator.tsx`

**Comment by gabriellaputri** [comment]
Friday, August 25, 2023

Synced changes to `alpha-v2` branch:
- Change the value of Expo's experience ID
- Update on the next version documentation
- Updating package `expo` version 40.0.6
- Refactoring refresh token implementation
- Updating packages `react-native` and `expo-updates`

**Comment by gabriellaputri** [comment]
Friday, September 1, 2023

Synced changes to `alpha-v2` branch:
- Adjust the condition to remove notification badge after clicking
- Fix the issue with navigation after clicking push notification
- Fix `undefined` image name after upload
- Setup `winston-daily-rotate-file`
- Fix the issue with attempting to create a post in "Uncategorized" category
- Fix Github action's warning by updating dependency

**Comment by gabriellaputri** [comment]
Thursday, September 14, 2023

Synced changes to `alpha-v2` branch:
- Delete everything related to `frequentPosterUserId`

**Comment by gabriellaputri** [comment]
Friday, September 15, 2023

Synced changes to `v2.0.0-beta` branch:
- Delete everything related to `frequentPosterUserId`
- Create a patch to fix warning from `react-navigation/stack`
- Reset navigation after create new post
